### PR TITLE
Feat combo layers

### PR DESCRIPTION
**Pull request by mcrosson**
Add layer filters to combos. This allows a combo to be active for one or more layers as well as global in scope.

The default is for a combo to be globally applicable.

### CODE REVIEW

**Comment by mcrosson** [comment]
@okke-formsma Would you be willing to review this?

File: app/src/combo.c [hunk 1]
```diff
@@ -105,16 +107,37 @@ static int initialize_combo(struct combo_cfg *new_combo) {
 }
 
 static int setup_candidates_for_first_keypress(int32_t position, int64_t timestamp) {
+    // Due to layer filtering we need to track the candidates[index] stuff by hand
+    int number_of_combo_candidates = -1; // First increment moves this to 0 as expected
+    uint8_t highest_active_layer = zmk_keymap_highest_layer_active();
     for (int i = 0; i < CONFIG_ZMK_COMBO_MAX_COMBOS_PER_KEY; i++) {
+        number_of_combo_candidates++; // Increment on each pass of the loop
```
Comment (okke-formsma) [inline app/src/combo.c:114]: I think we can clean this up a bit; 

initialize number_of_combo_candidates to 0, then remove this line and only increment it in the `if (combo_layer_valid)` clause.

File: app/src/combo.c [hunk 2]
```diff
@@ -105,16 +107,37 @@ static int initialize_combo(struct combo_cfg *new_combo) {
...
+        if (combo->layers[0] == -1) {
+            combo_layer_valid = true;
+        } else {
+            for (int j = 0; j < sizeof(combo->layers) / sizeof(combo->layers[0]); j++) {
+                if (combo->layers[j] == highest_active_layer) {
+                    combo_layer_valid = true;
+                    break;
+                }
+            }
+        }
```
Comment (okke-formsma) [inline app/src/combo.c:131]: I think we should extract lines 121-131 into a separate function like `combo_active_on_layer(combo, layer)`

File: docs/docs/features/combos.md [hunk 3]
```diff
@@ -29,6 +30,7 @@ Combos configured in your `.keymap` file, but are separate from the `keymap` nod
 - `key-positions` is an array of key positions. See the info section below about how to figure out the positions on your board.
 - `bindings` is the behavior that is activated when the behavior is pressed.
 - (advanced) you can specify `slow-release` if you want the combo binding to be released when all key-positions are released. The default is to release the combo as soon as any of the keys in the combo is released.
+- (advanced) you can specity a list of layers (`layers = <0 1...>`) to limit a combo to specific layers. this is an _optional_ parameter and defaults to `-1` which is global scope.
```
Comment (okke-formsma) [inline docs/docs/features/combos.md:33]: I think this should not be in the 'advanced' section but in the default.

**Review by okke-formsma — CHANGES_REQUESTED** [review]
In general, looks good. I see room for a few improvements.

Also please add at least one test to cover this functionality.

File: app/src/combo.c (hunk 1 above)
Comment (mcrosson) [reply app/src/combo.c]: Re-worked the incrementing logic to simplify/cleanup

File: app/src/combo.c (hunk 2 above)
Comment (mcrosson) [reply app/src/combo.c]: Extracted this as a method per suggestion

File: docs/docs/features/combos.md (hunk 3 above)
Comment (mcrosson) [reply docs/docs/features/combos.md]: Move up in docs to be the last item before advanced and removed advanced modifier from text

**Comment by mcrosson** [comment]
> Also please add at least one test to cover this functionality.

Test case added.


This should be good for re-review

File: app/dts/bindings/zmk,combos.yaml [hunk 4]
```diff
@@ -20,3 +20,5 @@ child-binding:
       default: 50
     slow-release:
       type: boolean
+    layers:
+      type: array
```
Comment (okke-formsma) [inline app/dts/bindings/zmk,combos.yaml:24]: I think setting a default of -1 would be nice.

File: app/src/combo.c [hunk 5]
```diff
@@ -104,17 +106,40 @@ static int initialize_combo(struct combo_cfg *new_combo) {
     return 0;
 }
 
+static bool combo_active_on_layer(struct combo_cfg *combo, uint8_t layer) {
```
Comment (okke-formsma) [inline app/src/combo.c:109]: maybe this is simpler?
```
static bool combo_active_on_layer(struct combo_cfg *combo, uint8_t layer) {
    if (combo->layers[0] == -1) {
        // -1 in the first layer position is global layer scope
        return true;
    }
    for (int j = 0; j < CONFIG_ZMK_COMBO_MAX_LAYER_FILTERS_PER_COMBO; j++) {
        if (combo->layers[j] == layer) {
            return true;
        }
    }
    return false;
}
```

File: app/src/combo.c [hunk 6]
```diff
@@ -104,17 +106,40 @@ static int initialize_combo(struct combo_cfg *new_combo) {
...
+                combo_layer_valid = true;
+                break;
+            }
+        }
+    }
+    return combo_layer_valid;
+}
+
 static int setup_candidates_for_first_keypress(int32_t position, int64_t timestamp) {
+    // Due to layer filtering we need to track the candidates[index] stuff by hand
```
Comment (okke-formsma) [inline app/src/combo.c:127]: I think we can drop the comment.

File: app/dts/bindings/zmk,combos.yaml (hunk 4 above)
Comment (okke-formsma) [reply app/dts/bindings/zmk,combos.yaml:24]: Hmmm the default is managed by `.layers = DT_PROP_OR(n, layers, {-1}),`. That works for me.

File: app/src/combo.c (hunk 5 above)
Comment (mcrosson) [reply app/src/combo.c:109]: Nice catch, this is far simpler

File: app/src/combo.c (hunk 6 above)
Comment (mcrosson) [reply app/src/combo.c:127]: Agreed

**Comment by okke-formsma** [comment]
I've found a bug

```
#include <dt-bindings/zmk/keys.h>
#include <behaviors.dtsi>
#include <dt-bindings/zmk/kscan-mock.h>

/* it is useful to set timeout to a large value when attaching a debugger. */
#define TIMEOUT (60*60*1000)

/ {
	combos {
		compatible = "zmk,combos";
		combo_one {
			timeout-ms = <TIMEOUT>;
			key-positions = <0 1>;
			bindings = <&kp X>;
			layers = <1>;
		};
	};

	keymap {
		compatible = "zmk,keymap";
		label ="Default keymap";

		default_layer {
			bindings = <
				&kp A &kp B
				&kp C &tog 1
			>;
		};
	};
};

&kscan {
	events = <
		/* Combo One */
		ZMK_MOCK_PRESS(0,0,10) 
		ZMK_MOCK_PRESS(0,1,10) 
		ZMK_MOCK_RELEASE(0,0,10) 
		ZMK_MOCK_RELEASE(0,1,10)
	>;
};
```

This presses the combo even though we're still on layer 0. Thishappens because the default value for the layers that are not specified is 0:

image

The way to fix this is to store the number of layers with the layers array. See `key_position_len` for an example of how to do this.

**Comment by mcrosson** [comment]
@okke-formsma Pushed a fix for the bug you found as well as other cleanup previously mentioned

File: app/src/combo.c [hunk 7]
```diff
@@ -451,6 +472,8 @@ ZMK_SUBSCRIPTION(combo, zmk_position_state_changed);
         .behavior = KEY_BINDING_TO_STRUCT(0, n),                                                   \
         .virtual_key_position = ZMK_KEYMAP_LEN + __COUNTER__,                                      \
         .slow_release = DT_PROP(n, slow_release),                                                  \
+        .layers = DT_PROP_OR(n, layers, {-1}),                                                     \
```
Comment (okke-formsma) [inline app/src/combo.c:475]: let's use DT_PROP() here without the default

**Review by okke-formsma — APPROVED** [review]
Thanks!

File: app/src/combo.c [hunk 8]
```diff
@@ -33,6 +34,8 @@ struct combo_cfg {
     // the virtual key position is a key position outside the range used by the keyboard.
     // it is necessary so hold-taps can uniquely identify a behavior.
     int32_t virtual_key_position;
+    int8_t layers[CONFIG_ZMK_COMBO_MAX_LAYER_FILTERS_PER_COMBO];
+    int32_t layers_len;
```
Comment (petejohanson) [inline app/src/combo.c:38]: Why use a fixed length array here, if we're also going to need to store a length parameter? We should be able to do something like we do in `app/drivers/kscan/kscan_gpio_direct.c` and make this:

```suggestion
    int32_t layers_len;
    int8_t layers[];
```

If we go this route, we can skip the fixed limit in the Kconfig you added.

File: app/Kconfig [hunk 9]
```diff
@@ -301,6 +301,10 @@ config ZMK_COMBO_MAX_KEYS_PER_COMBO
 	int "Maximum number of keys per combo"
 	default 4
 
+config ZMK_COMBO_MAX_LAYER_FILTERS_PER_COMBO
```
Comment (petejohanson) [inline app/Kconfig:304]: See my other comment, I think we can skip this.

**Review by petejohanson — CHANGES_REQUESTED** [review]
Just one minor comment. Otherwise looks good.

File: docs/docs/features/combos.md [hunk 10]
```diff
@@ -49,3 +51,4 @@ There are three global combo parameters which are set through KConfig. You can s
 - `CONFIG_ZMK_COMBO_MAX_PRESSED_COMBOS` is the number of combos that can be active at the same time. Default 4.
 - `CONFIG_ZMK_COMBO_MAX_COMBOS_PER_KEY` is the maximum number of combos that can be active on a key position. Defaults to 5. (So you can have 5 separate combos that use position `3` for example)
 - `CONFIG_ZMK_COMBO_MAX_KEYS_PER_COMBO` is the maximum number of keys that need to be pressed to activate a combo. Default 4. If you want a combo that triggers when pressing 5 keys, you'd set this to 5 for example.
+- `ZMK_COMBO_MAX_LAYER_FILTERS_PER_COMBO` is the maximum number of layer filters that can be specified. Default 4. If you want use layer filters with a combo that spans more than 4 layers, you'll need to change this value.
```
Comment (petejohanson) [inline docs/docs/features/combos.md:54]: This entry can be removed now that you 'nixed the Kconfig setting.

**Review by petejohanson — APPROVED** [review]

**Pull request merged**